keps/83-workload-preemption/README.md

```diff
@@ -0,0 +1,449 @@
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+### Increased admission latency
+
+## Design Details
+
+The proposal consists of new API fields and a preemption algorithm.
+
+### API
+
+#### Workload changes
+
+The new API fields in a Workload describe the policies and preferences for
+preempting the Workload.
+
+```golang
+
+const WorkloadStarted = "Started"
+
+type WorkloadSpec struct {
+  ...
+  // preemption describes the policies and preferences for preempting this
+  // Workload.
+  // preemption is immutable once set.
+  Preemption *WorkloadPreemption
+}
+
+type WorklodPreemption struct {
+  // disruptionCostMilestones defines a cost curve for preempting this workload
+  // based on the time the workload has been running.
+  // The curve is defined by points of (time, cost). The cost is calculated from
+  // a linear interpolation of the cost between the two time points that contain
+  // the current workload runtime.
+  // The first point is implicitly (0, 0)
+  // The cost from the last milestone is constant.
+  //
+  // For example: given the points (60, 0) and (120, 100), the cost of
+  // preempting the Workload after it has been running for:
+  // - 30s is 0,
+  // - 90s is 50,
+  // - 300s is 100.
+  //
+  // disruptionCostMilestones[i].time > disruptioncostMilestones[j].time for i > j
+  DisruptionCostMilestones []DisruptionCostMilestone.
+}
+
+type DisruptionCostMilestone struct {
+  // seconds is the time since the Workload started (timestamp of the Started
+  // condition on the Workload).
+  Seconds int64
+
+  // cost is an integer representation of the cost of disrupting the Workload
+  // at a given time. The value has meaning when compared to the cost of
+  // disrupting other Workloads.
+  //
+  // 0 <= cost <= 100
+  Cost int32
+}
+```
+
+#### LocalQueue changes
+
+The new API fields in LocalQueue describe the defaults that apply to Workloads
+in this LocalQueue.
+
+```golang
+type LocalQueueSpec struct {
+  ...
+  // preemptionDefaults are the defaults applied to Workloads assigned to this
+  // LocalQueue. The defauts are applied to Workloads at the time of the
+  // Workload admission.
+  PreemptionDefaults *WorkloadPreemption
+}
+```
+
+#### ClusterQueue changes
+
+The new API fields in ClusterQueue describe how to influence the selection
+of Workloads to preempt.
+
+```golang
+const ClusterQueuePendingWorkloadsUnderMinQuota = "PendingWorkloadsUnderMinQuota"
+
+type ClusterQueueSpec struct {
+  ...
+  // preemption describes policies to preempt Workloads from ClusterQueues
+  // borrowing resources when this ClusterQueue has pending Workloads and the
+  // min quota is unmet.
+  preemption ClusterQueuePreemption
+}
+
+type ClusterQueueStatus struct {
+  ...
+  conditions []metav1.Condition
+}
+
+type ClusterQueuePreemption struct {
```

Comment:
Explain if the two APIs are mutually exclusive, and if not, what is the behavior when both are set